Refactor cleanup task and add checks for gempak files

DavidHuber-NOAA opened the pull request:
# Description
This adds more specific checks for gempak files and makes an attempt at making the cleanup job less disk-intensive by running a single `find -f` command.

Resolves #4271 
Resolves #4166 
# Type of change
- [ ] Bug fix (fixes something broken)
- [x] New feature (adds functionality)
- [x] Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO
- Does this change require an update to any of the following submodules? NO

# How has this been tested?
Will test on an extended test.

# Checklist
- [ ] Any dependent changes have been merged and published
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have documented my code, including function, input, and output descriptions
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] This change is covered by an existing CI test or a new one has been added
- [ ] Any new scripts have been added to the .github/CODEOWNERS file with owners
- [ ] I have made corresponding changes to the system documentation if necessary

DavidHuber-NOAA committed:
Enable set -o pipefail to catch errors in piped commands

DavidHuber-NOAA committed:
Refactor cleanup task and add checks for gempak files

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/save_gempak_1p00

DavidHuber-NOAA committed:
Shellcheck fixes

DavidHuber-NOAA commented:
FYI @CatherineThomas-NOAA @JessicaMeixner-NOAA @aerorahul. I will test this on an extended case with spoofed older COM structures.

DavidHuber-NOAA committed:
Ignore 2312

DavidHuber-NOAA committed:
Remove unused variable

DavidHuber-NOAA committed:
Disable pipefail exits for now

DavidHuber-NOAA committed:
Shellcheck issues

DavidHuber-NOAA committed:
Merge branch 'feature/save_gempak_1p00' of github.com:davidhuber-noaa/global-workflow into feature/save_gempak_1p00

DavidHuber-NOAA committed:
Fix bufrsnd tarball creation

DavidHuber-NOAA commented on scripts/exgfs_atmos_postsnd.sh:
```diff
@@ -148,12 +148,11 @@ else
+# Tar and gzip the bufr files created placed in COM
+###################################################
+tar -czf "${RUN}.${cycle}.bufrsnd.tar.gz" -C "${COMOUT_ATMOS_BUFR}" bufr.*
+cpfs "${RUN}.${cycle}.bufrsnd.tar.gz" "${COMOUT_ATMOS_BUFR}/"
```
@BoCui-NOAA Does this look OK to you?

BoCui-NOAA replied:
Yes, that works for me. Thank you for taking care of it.

DavidHuber-NOAA committed:
Correct archiving to actually pull the data used

DavidHuber-NOAA committed:
Correct positional arguments

DavidHuber-NOAA committed:
Tar all bufr files

BoCui-NOAA commented on ush/gfs_sndp.sh:
```diff
@@ -19,18 +19,21 @@ cd $DATA/$m
+       bufr_target="bufr.${stn}.${PDY}${cyc}"
+       cpreq "${COMIN_ATMOS_BUFR}/${bufr_target}" "${DATA}/${m}/bufrin"
+       # Also write the filename to a list for files to tar
+       echo "${bufr_target}" >> "${DATA}/${m}/bufr_snd_files.list"
```
Do we still need this line to generate the filename list?

BoCui-NOAA replied:
@DavidHuber-NOAA Do you have a test so I can check the generated tar file? Thanks!

DavidHuber-NOAA replied:
@BoCui-NOAA Yes, you can find test case tar file here: `/lfs/h2/emc/eib/noscrub/david.huber/rt_gempak/COMROOT/C96_atm3DVar_extended_gempak/gfs.20211222/06/products/atmos/bufr/gfs.t06z.bufrsnd.tar.gz`

BoCui-NOAA replied:
@DavidHuber-NOAA The tar file looks fine to me. The only thing I noticed is the new setting for COMIN_ATMOS_BUFR, which points to the same directory as COMOUT_ATMOS_BUFR. Will this cause confusion?

DavidHuber-NOAA replied:
@BoCui-NOAA thanks for checking the tar file!  I don't think it will cause confusion as there are some other J-Jobs where we have a COMIN and COMOUT pointing to the same directory. This fulfills an EE2 requirement that all COM variables are labeled correctly for reads (COMIN) and writes (COMOUT).

BoCui-NOAA replied:
Thanks for the explanation. The updated scripts for the BUFR sounding part look okay to me.

DavidHuber-NOAA committed:
Use a find command instead

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/save_gempak_1p00

DavidHuber-NOAA committed:
Do not load gh anymore

DavidHuber-NOAA committed:
Enable archiving on WCOSS2 by default

DavidHuber-NOAA committed:
Make the bufr soundings and surface files EE2-compliant

DavidHuber-NOAA commented on ush/gfs_bfr2gpk.sh:
```diff
@@ -46,13 +49,13 @@ date
-   "${DBNROOT}/bin/dbn_alert" MODEL GFS_PTYP_SFC "${job}" "${COMOUT_ATMOS_GEMPAK}/${sfc}"
-   "${DBNROOT}/bin/dbn_alert" MODEL GFS_PTYP_SND "${job}" "${COMOUT_ATMOS_GEMPAK}/${snd}"
+   "${DBNROOT}/bin/dbn_alert" MODEL GFS_PTYP_SFC "${job}" "${COMOUT_ATMOS_GEMPAK}/${sfc_out}"
+   "${DBNROOT}/bin/dbn_alert" MODEL GFS_PTYP_SND "${job}" "${COMOUT_ATMOS_GEMPAK}/${snd_out}"
```
FYI @BoCui-NOAA

BoCui-NOAA replied:
@DavidHuber-NOAA The changes look good to me. When you have a test ready, please let me know so I can double-check. Thanks!

DavidHuber-NOAA replied:
@BoCui-NOAA sure thing, here are a pair:

```
/lfs/h2/emc/ptmp/david.huber/rt_gempak/COMROOT/C96_atm3DVar_extended_gempak/gfs.20211222/18/products/atmos/gempak/gfs_2021122218.soundings.bufr
/lfs/h2/emc/ptmp/david.huber/rt_gempak/COMROOT/C96_atm3DVar_extended_gempak/gfs.20211222/18/products/atmos/gempak/gfs_2021122218.sfc.bufr
```

BoCui-NOAA replied:
The two files are what we expected. Thanks!

DavidHuber-NOAA committed:
Update soundings tarball name to be EE2-compliant

DavidHuber-NOAA committed:
Correct atmos product dependencies

JessicaMeixner-NOAA commented on dev/parm/config/gfs/config.cleanup:
```diff
@@ -9,22 +9,30 @@ source "${EXPDIR}/config.resources" cleanup
+export SELECTIVE_CLEANUP_MAX=144
+export SELECTIVE_CLEANUP_MIN=24
+export GEMPAK_CLEANUP_MAX=240
+export RTOFS_CLEANUP_MAX=48
```
I think this could be smaller by default.

DavidHuber-NOAA replied:
@JessicaMeixner-NOAA Are you referring to `RTOFS_CLEANUP_MAX` or all of the `MAX` values? If the former, would `24` make sense?

JessicaMeixner-NOAA replied:
Just RTOFS_CLEANUP_MAX.

DavidHuber-NOAA replied:
Done.

DavidHuber-NOAA committed:
Replace tabs with spaces

DavidHuber-NOAA committed:
Increase memory for gempak jobs (needed for meta task)

DavidHuber-NOAA committed:
Make cleanup script smarter

Do selective cleanup by determining date ranges for files to keep. Only keep files based on the dates selected.

DavidHuber-NOAA committed:
Merge develop; fix conflicts

DavidHuber-NOAA commented:
I finished thorough testing of this PR via an extended C96 test case (out 18 days) to verify that cleanup is being handled correctly based on exclusion lists. The COM directory can be found here: `/lfs/h2/emc/ptmp/david.huber/rt_gempak/COMROOT/C96_atm3DVar_extended_gempak`.  Note that the experiment was spoofed after the 4th full cycle (from 2021122300 through 2022011000).  Opening for review.

DavidHuber-NOAA commented on dev/parm/config/gfs/config.cleanup:
```diff
@@ -9,22 +9,30 @@ source "${EXPDIR}/config.resources" cleanup
-    export RMOLDEND=222
+export SELECTIVE_CLEANUP_MAX=144
+export SELECTIVE_CLEANUP_MIN=24
+export GEMPAK_CLEANUP_MAX=240
```
@ChristopherHill-NOAA Can you verify that this is the correct setting for the `*gfs_1p00_*` files needed for the GEMPAK meta task?  This will delete the target files that are older than 240 hours before the current cycle at the end of the cycle (e.g. if the current cycle is 2022012000, delete `*gfs_1p00_*` files from cycles before 2022011000).

ChristopherHill-NOAA replied:
@DavidHuber-NOAA This is generally correct - only the `*gfs_1p00*` files are considered as being needed by the GEMPAK meta task as early as -240 hours before the active cycle, and such data older than -240 hours can be deleted.

From the `COMROOT/C96_atm3DVar_extended_gempak` test case, it is observed that only the `products/atmos/gempak/1p00` subdirectories and their content (i.e. `*gfs_1p00*` files) are retained within the cycle range of 2022010300 to 2022010718 -- which are -168 and -54 hours relative to the last cycle 2022011000 of the test case, respectively.  Before and after this cycle range, all (spoofed) subdirectories and files under `products/atmos` remain present.

ChristopherHill-NOAA replied:
It is supposed that the desired effect is a) to retain only `*gfs_1p00*` files within the range of `{GEMPAK_CLEANUP_MAX}` to `{SELECTIVE_CLEANUP_MIN}` and b) to delete those `*gfs_1p00*` files older than`{GEMPAK_CLEANUP_MAX}`, since the value of `GEMPAK_CLEANUP_MAX` is assumed as being greater than the value of `SELECTIVE_CLEANUP_MAX`.  It is further observed that the `products/atmos/gempak/1p00/gfs_1p00*` data content remains present in directories representing cycles earlier than the value of `{GEMPAK_CLEANUP_MAX}`.  It is noted that the `SELECTIVE_CLEANUP_{MIN/MAX}` variables are intended only to focus file cleanup over a limited range of cycles.

DavidHuber-NOAA replied:
@ChristopherHill-NOAA Thank you for checking the directories.  I had created a bug during a merge after testing and did not thoroughly test again. I have now fixed that bug and further improved the scripting. Would you mind checking the COM directory again?

For clarity:

- My test case runs from cycle 2021122018 through 2022011000. I have spoofed data from 2021122200 through 2022011000.
- I am only running my test on the last cycle (2022011000).
- I am working on the assumption that when files are needed for the GEMPAK meta task, 240 hours (40 cycles) worth of `products/atmos/gempak/1p00/gfs_1p00*` files are required (i.e. back to cycle `2021123100`).
- `prepbufr`, `cnvstat.tar`, and `analysis.atm.a*.nc` files should be retained for 144 hours (24 cycles, or back to `2022010400`)
- The cleanup task starts looking in directories 24 hours older than the maximum retention period (240 hours in the `RUN=gfs` case) meaning 264 hours before 2022011000, or cycle 2021123000.
- Directories older than 240 hours (i.e. `gfs.20211230/00`, `gfs.20211230/06`, `gfs.20211230/12`, and `gfs.20211230/18`) should be deleted
- Empty `RUN.PDY` directories (i.e. `gfs.20211230`) should also be deleted

Please correct me if I am wrong about any of these bullets.  If everything looks correct, I will run another test for cycle `2022011006` to verify that the cleanup increments by a cycle for all excepted files.

ChristopherHill-NOAA replied:
The bullet points appear to be correct, and representative of the intended effort to clean all but `*gfs_1p00*` files out to -240 hours and to clean those remaining `*gfs_1p00*` files beyond -240 hours.

From the latest test, it is observed that the `gfs.20211230` directory is missing, and assumed to have been deleted by the cleanup script.  The `*gfs_1p00*` data remain for cycles 2021123100 through 2022010900, with the directories representing cycles 2021122918 and older being unaffected.

Some of the `*gfs_1p00*` files are unexpectedly missing under each of the <ins>cycle 18</ins> directories.  Specifically missing from `gfs.yyyymmdd/18/products/atmos/gempak/1p00` are `*gfs_1p00*` files with the following suffixes:
`   f003, f015, f027, f033, f039, f042, f045, f051, f054`
These files may have been inadvertently deleted, or were not captured in the spoofing scheme.

DavidHuber-NOAA replied:
@ChristopherHill-NOAA I believe this was caused by the ptmp scrubber. I have rerun the test case from the beginning, i.e.

- starting at 2021122018 and ending on 2022011000
- spoofed data from 2021122200 through 2022011000
    - The filenames were mimicked by inventorying the contents of gfs.20211221 and gdas.20211221
- The gdas_cleanup and gfs_cleanup jobs were then run on 2022011000
    - This resulted in the removal of all GFS cycles on 20211230 and all GDAS cycles on 20220103 from the COM directory
- Note that the forecasts now go out to 384 hours instead of 240

I then verified that the `*gfs_1p00*` files were present at/after cycle 2021123100 at all forecast hours.

You can find the COM directory here: `/lfs/h2/emc/ptmp/david.huber/rt_gempak3/COMROOT/C96_atm3DVar_extended_gempak3`
and the associated EXPDIR here: `/lfs/h2/emc/ptmp/david.huber/rt_gempak3/EXPDIR/C96_atm3DVar_extended_gempak3`

Does this look good to you?

ChristopherHill-NOAA replied:
@DavidHuber-NOAA  The latest workflow run - with modified GEMPAK file cleanup through case `C96_atm3DVar_extended_gempak3` - appears to provide the desired results.  Each of the expected files is confirmed to present in the `gfs.yyyymmdd/cc/products/atmos/gempak/1p00` directories over the cycle range of 2021123100 to 2022010900 following invocation of `exglobal_cleanup.sh` with `GEMPAK_CLEANUP=240`.

DavidHuber-NOAA replied:
Great, thank you for confirming @ChristopherHill-NOAA!

DavidHuber-NOAA committed:
Reduce RTOFS retention to 24 hours

CatherineThomas-NOAA commented on dev/parm/config/gfs/config.cleanup:
```diff
@@ -9,22 +9,30 @@ source "${EXPDIR}/config.resources" cleanup
 #--starting and ending hours of previous cycles to be removed from rotating directory
-export RMOLDSTD=144
-export RMOLDEND=24
+# Selectively remove files between SELECTIVE_RM_MIN and SELECTIVE_RM_MIN hours old, based on exclude_string
```
Nitpick: This comment should be updated to be consistent with change in variable name `SELECTIVE_RM*` --> `SELECTIVE_CLEANUP*`

DavidHuber-NOAA replied:
Good catch.  I've corrected the comments and renamed the variables throughout to a clearer convention.

CatherineThomas-NOAA commented on scripts/exglobal_cleanup.sh:
```diff
@@ -15,15 +17,51 @@ if [[ "${CLEANUP_COM:-YES}" == NO ]]; then
-last_rtofs=$(date --utc +%Y%m%d%H -d "${PDY} ${cyc} -${RMOLDRTOFS:-48} hours")
+first_selective_date=$(date --utc +%Y%m%d%H -d "${PDY} ${cyc} -${SELECTIVE_CLEANUP_MAX} hours")
+last_selective_date=$(date --utc +%Y%m%d%H -d "${PDY} ${cyc} -${SELECTIVE_CLEANUP_MIN} hours")
+last_rtofs_date=$(date --utc +%Y%m%d%H -d "${PDY} ${cyc} -${RTOFS_CLEANUP_MAX} hours")
```
Should this be `first_rtofs_date` to be consistent with the other `MAX` variables?

DavidHuber-NOAA replied:
Yes, though I have changed this to a `rtofs_cutoff_date` to make it clearer what it represents.

CatherineThomas-NOAA commented:
@DavidHuber-NOAA - Thanks for taking care of this!  It will really reduce our disk usage while retaining what we need for gempak in the realtime.

I'm not fully sure how you set up and ran the extended test that you included, but we would expect that the older gdas and gfs directories would NOT be fully cleaned out, correct?  For example, `gdas.20211230/00/analysis/atmos` is still fully populated, but `gdas.20220108/00/analysis/atmos` is cleaned, which is consistent if you skipped the cycles and only ran the cleanup at the latest date (since the cleanup job only looks back so far).

DavidHuber-NOAA committed:
Remove extra }

DavidHuber-NOAA committed:
Rename RTOFS and GEMPAK variables

DavidHuber-NOAA committed:
Clarify RTOFS cleanup language

DavidHuber-NOAA committed:
Add DO_BUFRSND dependency on gempak/bufr sounding outputs

DavidHuber-NOAA committed:
Linter cleanup

DavidHuber-NOAA committed:
Rename variables to more meaningful names

DavidHuber-NOAA committed:
Only delete empty RUN.PDY directories

DavidHuber-NOAA committed:
Add check for invalid GEMPAK retention

DavidHuber-NOAA commented:
> @DavidHuber-NOAA - Thanks for taking care of this! It will really reduce our disk usage while retaining what we need for gempak in the realtime.
> 
> I'm not fully sure how you set up and ran the extended test that you included, but we would expect that the older gdas and gfs directories would NOT be fully cleaned out, correct? For example, `gdas.20211230/00/analysis/atmos` is still fully populated, but `gdas.20220108/00/analysis/atmos` is cleaned, which is consistent if you skipped the cycles and only ran the cleanup at the latest date (since the cleanup job only looks back so far).

@CatherineThomas-NOAA I only ran this for the `2022011000` cycle and, by design, it only looks back a further 24 hours from `max(CLEANUP_SELECTIVE_MAX, CLEANUP_GEMPAK)`, so it only looks back to cycle `2021123000`. Since this did not run for any other cycle, older directories stay in tact.

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/save_gempak_1p00

JessicaMeixner-NOAA commented:
@DavidHuber-NOAA  - I've looked things over and don't see anything else, but @CatherineThomas-NOAA  has been much more in the weeds with several parts of this than  I have so I'm going to let her do the more careful review of this.  Thank you so much for this PR!

CatherineThomas-NOAA commented:
Thanks for addressing my comments @DavidHuber-NOAA.  The only other missing piece I see is @ChristopherHill-NOAA's comment about the 18z files:

> Some of the *gfs_1p00* files are unexpectedly missing under each of the cycle 18 directories. Specifically missing from gfs.yyyymmdd/18/products/atmos/gempak/1p00 are *gfs_1p00* files with the following suffixes:
   f003, f015, f027, f033, f039, f042, f045, f051, f054

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/save_gempak_1p00

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/save_gempak_1p00

DavidHuber-NOAA committed:
Re-disable HPSS archiving on WCOSS2 by default

DavidHuber-NOAA committed:
Disable C48_S2SW on WCOSS2

This test is covered by C48_S2SW_extended

DavidHuber-NOAA committed:
Turn on HPSS archiving for extended tests

DavidHuber-NOAA commented:
@CatherineThomas-NOAA @aerorahul Could you please review this PR?

aerorahul commented on dev/ci/cases/pr/C48_S2SW.yaml:
```diff
@@ -11,6 +11,9 @@ experiment:
   yaml: {{ HOMEgfs }}/dev/ci/cases/yamls/gfs_defaults_ci.yaml
 
+skip_ci_on_hosts:
+  - wcoss2
```
why are we skipping this test on WCOSS2?

DavidHuber-NOAA replied:
Because we run the `C48_S2SW_extended` test. It's a low-weight test, so I'm fine with keeping it if desired, but I thought we could at least save some compute hours.

aerorahul replied:
yeah. I am fine with skipping this lite test on WCOSS2 in favor of the extended test running the products.

aerorahul commented on parm/archive/gfs_downstream.yaml.j2:
```diff
@@ -3,7 +3,7 @@ gfs_downstream:
     target: "{{ ATARDIR }}/{{ cycle_YMDH }}/gfs_downstream.tar"
     required:
-        {% if DO_GEMPAK %}
+        {% if DO_GEMPAK and DO_BUFRSND %}
```
Should this block be separated out into 2; one each for GEMPAK and BUFRSND, in case one is ON and other is OFF?

DavidHuber-NOAA replied:
That's a good question.  I'm not sure if the `sfc.bufr` file is generated if `DO_GEMPAK` is enabled but `DO_BUFRSND` is not. Checking...

DavidHuber-NOAA replied:
@aerorahul No, both of these files only generate if both `DO_GEMPAK` and `DO_POSTSND` are on, so it should be an `and`.

aerorahul replied:
excellent.  Just checking.  Thanks for confirming.

aerorahul commented on ush/gfs_sndp.sh:
```diff
@@ -19,18 +19,19 @@ cd $DATA/$m
     do
-       cpreq "${COMOUT_ATMOS_BUFR}/bufr.${stn}.${PDY}${cyc}" "${DATA}/${m}/bufrin"
+       bufr_target="bufr.${stn}.${PDY}${cyc}"
+       cpreq "${COMIN_ATMOS_BUFR}/${bufr_target}" "${DATA}/${m}/bufrin"
```
Shouldn't this be `bufr_source` (if it needs to be a name) since it is the source for the copy?

DavidHuber-NOAA replied:
It doesn't need a name. This is a relic of an attempt I made to keep track of the input `bufr` files for the sounding jobs. I found a better way.  I'll revert 33-4.

DavidHuber-NOAA replied:
Reverted.

ChristopherHill-NOAA reviewed (commented):
The latest test of the workflow - specifically modifying GEMPAK data file cleanup through case `C96_atm3DVar_extended_gempak3` - provided the desired reduction in the retention of past cycle GEMPAK data files while retaining those GEMPAK files needed by `JGFS_ATMOS_GEMPAK_META`, thereby providing resolution to #4166.
 
With `DO_GEMPAK=YES`, `DO_BUFRSND=YES`, and `DO_ARCHCOM=YES` each set in `config.base` with the test case, and with no errors resulting after changes to `gfs_bfr2gpk.sh` and `gfs_sndp.sh`, there appears to be resolution to #4271.

DavidHuber-NOAA committed:
Revert the use of 'bufr_target'

DavidHuber-NOAA committed:
Remove now-unused variable

aerorahul reviewed (dismissed):
thanks for addressing the comments. lgtm.

DavidHuber-NOAA commented:
Launching CI on WCOSS2.

CatherineThomas-NOAA commented on dev/parm/config/gfs/config.cleanup:
```diff
@@ -9,22 +9,40 @@ source "${EXPDIR}/config.resources" cleanup
+fi
+
+# GEMPAK_CLEANUP must not be less than SELECTIVE_CLEANUP_MIN, which acts as the starting point for cleanup
+if [[ "${DO_GEMPAK}" == "YES" && ${GEMPAK_CLEANUP} -lt ${GEMPAK_CLEANUP} ]]; then
```
I think one of these `GEMPAK_CLEANUP`s is supposed to be `SELECTIVE_CLEANUP_MIN`.

DavidHuber-NOAA replied:
Yes, you are correct.  Fixed.  I will add this into the running test to verify it is correct now.

DavidHuber-NOAA committed:
Correct GEMPAK_CLEANUP check

CatherineThomas-NOAA reviewed (approved):
Looks great. Thanks @DavidHuber-NOAA!

DavidHuber-NOAA commented:
All tests passed on WCOSS2 including archiving for `C48_S2SW_extended` and `C96_atm3DVar_extended`.  Merging.
